readme.md

```diff
@@ -0,0 +1,148 @@
+
+## Installation
+
+```
+yarn add @kaliber/forms
+```
+
+## Usage
+
+Please look at the example for more advanced use-cases.
+
+```jsx
+import { useForm, useFormField, required, email } from '@kaliber/forms'
+
+const validationErrors = {
+  required: 'This field is required',
+  email: 'This is not a valid e-mail',
+}
+
+export function Simple() {
+  const { form: { fields }, submit } = useForm({
+    initialValues: {
+      name: '',
+      email: '',
+    },
+    fields: {
+      name: required,
+      email: [required, email],
+    },
+    onSubmit: handleSubmit,
+  })
+
+  return (
+    <form onSubmit={submit}>
+      <TextInput label='Name' field={fields.name} />
+      <TextInput label='Email' field={fields.email} />
+      <button type='submit'>Submit</button>
+    </form>
+  )
+
+  function handleSubmit(snapshot) {
+    console.log(snapshot)
+  }
+}
+
+function TextInput({ label, field }) {
+  const { name, state, eventHandlers } = useFormField(field)
+  const { value, error, showError } = state
+  return (
+    <>
+      <div>
+        <label htmlFor={name}>{label}</label>
+        <input
+          id={name}
+          type='text'
+          value={value === undefined ? '' : value}
+          {...eventHandlers}
+        />
+      </div>
+      {showError && <p>{validationErrors[error.id]}</p>}
+    </>
+  )
+}
+```
+![](https://media.giphy.com/media/c21EAVffi7zd6/giphy.gif)
+
+## Reference
+
+### useForm
+
+Defines a form.
+
+```js
+const {
+  form, // 'object' field containing the form
+  submit, // handler that can be used to submit the form
+  reset, // handler that can be used to reset the form
+} = useForm({
+  fields, // form structure
+  initialValues, // (optional) initial form values
+  validate, // (optional) validation for the complete form
+  handleSubmit, // called when the form is submitted
+})
+```
+
+|                |                                                                               |
+|----------------|-------------------------------------------------------------------------------|
+|`fields`        | An object with the shape: `{ [name: string]: FormField }`.|
+|`initialValues` | An object with the shape: `{ [name: keyof fields]: ValueFor<fields[name]> }`|
+|`validate`      | One of `Validate` or `Array<Validate>`|
+|`handleSubmit`  | A function that accepts a `Snapshot`|
+|                |                                                                               |
+|`FormField`     | One of `BasicField`, `ArrayField` or `ObjectField`|
+|`BasicField`    | One of `null`, `Validate` or `Array<Validate>`|
+|`ArrayField`    | Created with `array(fields)` or `array(validate, fields)`|
+|`ObjectField`   | Created with `object(fields)` or `object(validate, fields)`|
+|                |                                                                               |
+|`Validate`      | A function with the following shape: `(x, { form, context }) => falsy | { id, params }|`
```

Comment:
Wat houdt context precies in? Klopt het dat ik dat in de examples niet terug zie? Is het meer iets voor intern gebruik?

Wat wil de `| { id, params }|` precies zeggen, of gaat daar iets mis in de formatting?

Reply:
Ahh, goed gevonden, dat is een foutje. Zou `{ form, parents }` moeten zijn en dat is de `context`.